src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java

```diff
@@ -18,7 +26,231 @@
+        {
+            transform.add("origin", writeVec3(rootTransform.origin));
+        }
+
+        // If there were any transform properties set, add them to the output.
+        if (transform.size() > 0)
+        {
+            json.add("transform", transform);
+        }
+
+        return json;
+    }
+
+    private JsonArray writeVec3(Vector3f vector)
+    {
+        JsonArray array = new JsonArray();
+        array.add(vector.x());
+        array.add(vector.y());
+        array.add(vector.z());
+        return array;
+    }
+
+    private JsonArray writeQuaternion(Quaternion quaternion)
+    {
+        JsonArray array = new JsonArray();
+        array.add(quaternion.i());
+        array.add(quaternion.j());
+        array.add(quaternion.k());
+        array.add(quaternion.r());
+        return array;
+    }
+
+    public class RootTransformBuilder
+    {
+        private Vector3f translation = Vector3f.ZERO;
+        private Quaternion leftRotation = Quaternion.ONE.copy();
+        private Quaternion rightRotation = Quaternion.ONE.copy();
+        private Vector3f scale = new Vector3f(1, 1, 1);
+
+        private @Nullable String originName;
+        private @Nullable Vector3f origin;
+
+        /**
+         * Sets the translation of the root transform.
+         *
+         * @param translation the translation
+         * @return this builder
+         * @throws NullPointerException if {@code translation} is {@code null}
+         */
+        public RootTransformBuilder translation(Vector3f translation)
+        {
+            Preconditions.checkNotNull(translation, "Translation must not be null");
+            this.translation = translation;
+            return this;
+        }
+
+        /**
+         * Sets the left rotation of the root transform.
+         *
+         * @param rotation the left rotation
+         * @return this builder
+         * @throws NullPointerException if {@code rotation} is {@code null}
+         */
+        public RootTransformBuilder rotation(Quaternion rotation)
+        {
+            Preconditions.checkNotNull(rotation, "Rotation must not be null");
+            this.leftRotation = rotation;
+            return this;
+        }
+
+        /**
+         * Sets the left rotation of the root transform.
+         *
+         * @param leftRotation the left rotation
+         * @return this builder
+         * @throws NullPointerException if {@code leftRotation} is {@code null}
+         */
+        public RootTransformBuilder leftRotation(Quaternion leftRotation)
+        {
+            return rotation(leftRotation);
+        }
+
+        /**
+         * Sets the right rotation of the root transform.
+         *
+         * @param rightRotation the right rotation
+         * @return this builder
+         * @throws NullPointerException if {@code rightRotation} is {@code null}
+         */
+        public RootTransformBuilder rightRotation(Quaternion rightRotation)
```

Comment:
Should we also provide a `postRotation(Quaternion)` that calls to this method, to mimic the pair of `rotation` and `leftRotation`?